src/api_server/src/lib.rs

```diff
@@ -74,23 +84,92 @@ pub struct ApiServer {
-    ) -> Result<Self> {
-        Ok(ApiServer {
+    ) -> Self {
+        ApiServer {
             mmds_info,
             instance_info,
             api_request_sender,
             vmm_response_receiver,
             to_vmm_fd,
             vmm_fatal_error: false,
-        })
+        }
     }
 
+    /// Starts the HTTP Server by binding to the socket path provided as
+    /// an argument.
+    ///
+    /// # Arguments
+    ///
+    /// * `path` - the socket path on which the server will wait for requests.
+    /// * `start_time_us` - the timestamp for when the process was started in us.
+    /// * `start_time_cpu_us` - the timestamp for when the process was started in CPU us.
+    /// * `seccomp_filter` - the seccomp filter to apply.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use api_server::ApiServer;
+    /// use mmds::MMDS;
+    /// use seccomp::SeccompFilter;
+    /// use std::{
+    ///     convert::TryInto, io::Read, io::Write, os::unix::net::UnixStream, path::PathBuf,
+    ///     sync::mpsc::channel, thread, time::Duration,
+    /// };
+    /// use utils::{eventfd::EventFd, tempfile::TempFile};
+    /// use vmm::vmm_config::instance_info::InstanceInfo;
+    ///
+    /// let mut tmp_socket = TempFile::new().unwrap();
+    /// tmp_socket.remove().unwrap();
+    /// let path_to_socket = tmp_socket.as_path().to_str().unwrap().to_owned();
+    /// let api_thread_path_to_socket = path_to_socket.clone();
+    /// let instance_info = InstanceInfo {
+    ///     state: "Not started".to_string(),
+    ///     id: "test_serve_action_req".to_string(),
+    ///     vmm_version: "version 0.1.0".to_string(),
+    ///     app_name: "app name".to_string(),
+    /// };
+    ///
+    /// let to_vmm_fd = EventFd::new(libc::EFD_NONBLOCK).unwrap();
+    /// let (api_request_sender, _from_api) = channel();
+    /// let (to_api, vmm_response_receiver) = channel();
+    /// let mmds_info = MMDS.clone();
+    ///
+    /// thread::Builder::new()
+    ///     .name("fc_api_test".to_owned())
+    ///     .spawn(move || {
+    ///         ApiServer::new(
+    ///             mmds_info,
+    ///             instance_info,
+    ///             api_request_sender,
+    ///             vmm_response_receiver,
+    ///             to_vmm_fd,
+    ///         )
+    ///         .bind_and_run(
+    ///             PathBuf::from(api_thread_path_to_socket),
+    ///             Some(1),
+    ///             Some(1),
+    ///             SeccompFilter::empty().try_into().unwrap(),
+    ///         )
+    ///         .unwrap();
+    ///     })
+    ///     .unwrap();
+    ///
+    /// thread::sleep(Duration::new(0, 10_000_000));
+    /// let mut sock = UnixStream::connect(PathBuf::from(path_to_socket)).unwrap();
+    /// // Send a GET instance-info request.
+    /// assert!(sock.write_all(b"GET / HTTP/1.1\r\n\r\n").is_ok());
+    /// let mut buf: [u8; 100] = [0; 100];
+    /// assert!(sock.read(&mut buf[..]).unwrap() > 0);
+    /// ```
```

Comment:
Is there a newline between the code example and the next function? I don't know if the GitHub editor is showing it correctly.

Reply:
There is not a newline. Should there be?

Reply:
I think there should be a newline but it's a nit.